pymc_marketing/mmm/lift_test.py

```diff
@@ -0,0 +1,531 @@
+"""Lift test functions for the MMM."""
+
+from functools import partial
+from typing import Callable, Optional, Union
+
+import numpy as np
+import pandas as pd
+import pymc as pm
+import pytensor.tensor as pt
+from numpy import typing as npt
+
+from pymc_marketing.mmm.transformers import logistic_saturation
+from pymc_marketing.mmm.utils import michaelis_menten
+
+
+class MissingLiftTestError(Exception):
+    def __init__(self, missing_values: npt.NDArray[np.int_]) -> None:
+        self.missing_values = missing_values
+        super().__init__(
+            f"Some lift test values are not in the model: {missing_values}"
+        )
+
+
+Index = npt.NDArray[np.int_]
+Indices = dict[str, Index]
+Values = Union[npt.NDArray[np.int_], npt.NDArray[np.float_], npt.NDArray[np.str_]]
+
+
+def _lift_test_index(lift_values: Values, model_values: Values) -> Index:
+    # TODO: better support for datetime64 required for date coordinates
+    same_value = lift_values[:, None] == model_values
+    if not (same_value.sum(axis=1) == 1).all():
+        missing_values = np.argwhere(same_value.sum(axis=1) == 0).flatten()
+        raise MissingLiftTestError(missing_values)
+
+    return np.argmax(same_value, axis=1)
+
+
+def lift_test_indices(df_lift_test: pd.DataFrame, model: pm.Model) -> Indices:
+    """Get the indices of the lift test results in the model.
+
+    Assumes any column in the DataFrame is a coordinate in the model with the
+    same name.
+
+    Parameters
+    ----------
+    df_lift_test : pd.DataFrame
+        DataFrame with lift test results.
+    model : pm.Model
+        PyMC model with all the coordinates in the DataFrame.
+
+    Returns
+    -------
+    dict[str, np.ndarray]
+        Dictionary of indices for the lift test results in the model.
+
+    Raises
+    ------
+    MissingLiftTestError
+        If some lift test values are not in the model.
+
+    """
+
+    columns = df_lift_test.columns.tolist()
+    return {
+        col: _lift_test_index(df_lift_test[col].to_numpy(), np.array(model.coords[col]))
+        for col in columns
+    }
+
+
+def empirical_lift_measurements(
+    x_before: npt.NDArray[np.float_],
+    x_after: npt.NDArray[np.float_],
+    saturation_curve: Callable[[npt.NDArray[np.float_]], npt.NDArray[np.float_]],
+    pt=pt,
+) -> npt.NDArray[np.float_]:
+    """Calculate the empirical lift measurements at two spends.
```

Comment:
Again, the empirical lift was calculated (for example in CausalPy) when the user curated the lift test result. Here, we rather estimate the "theoretical lift given model". I don't know what a better function name would be. Maybe `estimate_modeled_lift`.